CHANGELOG.md

```diff
@@ -49,6 +49,12 @@ you should edit it (`kubectl edit validatingwebhookconfiguration
 kong-validations`) and add `kongclusterplugins` under the `resources` block for
 the `configuration.konghq.com` API group.
 
+#### Breaking Changes
+
+- The `--leader-elect` flag has been removed. Leader election is now enabled or
+  disabled automatically based on the database mode.
+  [#2053](https://github.com/Kong/kubernetes-ingress-controller/issues/2053)
+
```

Comment:
Do we want the same for 2.1.x? The commit is separate for that reason, so that we can cherry-pick only parts of this into 2.0.x if desired. It may make sense to put that in a separate PR if we want it gone in 2.1 just to simplify rebase/squash/whatever at the end of this one.

Simply not removing the flag in 2.0.x would still be breaking since we no longer honor the flag, but I could rework that to separate the logic that sets the mode automatically, so that for 2.0.x you need to enable leader election manually but it doesn't delete configuration if you do.

Reply:
> Do we want the same for 2.1.x? The commit is separate for that reason, so that we can cherry-pick only parts of this into 2.0.x if desired. It may make sense to put that in a separate PR if we want it gone in 2.1 just to simplify rebase/squash/whatever at the end of this one.
> Simply not removing the flag in 2.0.x would still be breaking since we no longer honor the flag, but I could rework that to separate the logic that sets the mode automatically, so that for 2.0.x you need to enable leader election manually but it doesn't delete configuration if you do.

It's a tricky one because we don't want to make a technically breaking change, but the underlining here is that we know if anyone explicitly set the flag and that setting was opposed to what the automatic configuration would have chosen it would result in a broken deployment.

In other words if there's anyone out there currently using the flag we have certainty they can only be setting it the exact way our new automation would set it, so not honoring the setting doesn't break anything because it will result in the same setting. The only thing we really need to avoid is dropping the flag entirely, because then those same theoretical users would get an upgrade failure.

For that reason I'm leaning towards disabling the functionality of the flag (because we wont be changing the actual value since it _has_ to be a certain way) but also marking the flag as deprecated and removing it in `v3` so that we're actually following semver and protecting that upgrade path.

Reply:
IMO we remove this one immediately:

- We do not intend to honor it at all. At most the deprecated flag would print a warning log if not set to the default.-
- I expect that usage of this flag in the wild is low: it wasn't present in 1.x and users that upgraded were accustomed to it being handled automatically, and we didn't add documentation indicating you should enable it in DB mode AFAIK. If people were using it, I expect we'd have received a report about it deleting configuration when enabled, and we didn't.
- Upgrade burden is pretty small. If you weren't using it, you do nothing. If you were, you'll see your upgrade fail and print a clear "unknown flag" error, and you run a new upgrade that removes it.

Had it been around longer I'd be less inclined to remove it quickly, but as 2.0.x is the only version that includes it, I figure we remove it now and make it clear it's gone to anyone that set it and didn't read the changelog.

Reply:
We use semver as a contract with our end-users that states we wont make backwards-incompatible changes except in major release versions. We assume, but we don't actually _know_ that nobody is using this flag explicitly and for all we know there's an example floating around on the internet somewhere that people are using which has this flag in it. Semver as a precept can often be used _instead of_ such telemetry: you can't always know what to expect of your end-users, but your end-users should be able to know what to expect of you. Given that the actual value of this configuration is inevitably going to match our automation, and given how small of a burden it would be on us to add a warning and some documentation and then mark this deprecated until `v3` where we can remove the few deprecated flags we have, I don't feel persuaded towards the position you're advocating for. However given how small this does feel in the grand scheme of things, and given that my perspective on this is to my knowledge and after reviewing our documentation _not_ something that I can see we ever unanimously agreed to institutionalize I also can't justify a hard block. As such I've dismissed my block on the PR and I invite @mflendrich to weigh in.

Reply:
I strongly agree with @shaneutt's view quoted below:

> Given that the actual value of this configuration is inevitably going to match our automation, and given how small of a burden it would be on us to add a warning and some documentation and then mark this deprecated until v3 where we can remove the few deprecated flags we have, I don't feel persuaded towards the position you're advocating for.

> Simply not removing the flag in 2.0.x would still be breaking since we no longer honor the flag

We're again in the fuzzy area of how "breaking" this change is. Given the below:

- _the actual value of this configuration is inevitably going to match our automation_ as @shaneutt pointed out
- we're viewing this as a fix, and we're reasonably confident that there exists no scenario where the change we're making will have a negative impact on anyone's configuration (at least that's the impression I have)

I'm tempted to think that leaving the flag there, but without any effect (other than printing a deprecation warning, I suppose) can get the root issue resolved with apparent full backwards-compatibility.,

Therefore, my call would be to leave the flag there, mark it as deprecated, and print a deprecation warning when one uses it.

Reply:
The flag is now deprecated instead: https://github.com/Kong/kubernetes-ingress-controller/pull/2053/commits/a71c15d5883c870536eb7b702b8e4da37dad864b